Migration of flatfield and pedestal components from lstchain

TjarkMiener opened the pull request:
This PR draft adds a migration proposal for the flatfield and pedestal components from lstchain into ctapipe. Shared functionality is moved into a parent component, while the different subcomponents hold the _process() function for the conventional processing of flatfield and pedestal events. Additional subcomponents with complex processing routines of flatfield and pedestal events can be added in the future.

Hckjs commented on src/ctapipe/calib/camera/extractor.py:
```diff
@@ -0,0 +1,426 @@
+            "sample_time": (
+                trigger_times[0] + (trigger_times[-1] - trigger_times[0]) / 2
+            ).unix
+            * u.s,
```
I think you need at least one of sample_time_{min, max} as timestamps to match the current event (trigger_times) with the related set of calibration coefficients.

maxnoe commented:
Hi @TjarkMiener thanks, this is much needed.

However, I think we should first discuss a bit how we want the general API of these classes to look like.

I think we certainly don't want to keep the current one.

* Do we need an abstract base class for the calibration like we have currently?

    This would depend on whether we can use the same general structure and algorithm per telescope. Will all telescopes use the F-Factor method?

* Do we use a custom tool to compute these calibration coefficients or do we integrate into `ctapipe-process`.

    I think specific tool is the way to go.

* What structure do we need? 

    Currently we have three classes, one for pedestal events, one for flatfield events and one tying these two together to compute the final calibration coefficients. That makes sense I think, but there is also a lot of duplication between the flatfield and the pedestal base class in lstchain currently. This should be refactored.

* How do we store the output?

* How do we read it back in for calibration?

There are a couple of related open issues:

* https://github.com/cta-observatory/ctapipe/issues/1836 (as  calibration events come with two gains)
* https://github.com/cta-observatory/ctapipe/issues/1397
* https://github.com/cta-observatory/ctapipe/issues/1226
* https://github.com/cta-observatory/ctapipe/issues/1147
* https://github.com/cta-observatory/ctapipe/issues/856

maxnoe commented on src/ctapipe/calib/camera/extractor.py:
```diff
@@ -0,0 +1,426 @@
+    sample_size : int
+        Minimal number of calibration events requested for the statistics
+    update_frequency : int
+        Calculation frequency of the statistics
```
This value is not clear to me from the explanation given.

FrancaCassol replied:
I agree, this is not clear

maxnoe commented on src/ctapipe/calib/camera/extractor.py:
```diff
@@ -0,0 +1,426 @@
+        self.broken_pixels = []  # masked pixels per event in sample
+
+        # load the waveform charge extractor
+        self.extractor = ImageExtractor.from_name(
```
Since you have `image_extractor` as a `TelescopeParameter`, you need to create all types requested, not just the one type.

See e.g.:
https://github.com/cta-observatory/ctapipe/blob/eb43957c838bfe30fd1a6403601b724b984a25e4/src/ctapipe/calib/camera/calibrator.py#L130

maxnoe commented on src/ctapipe/calib/camera/extractor.py:
```diff
@@ -0,0 +1,426 @@
+        return {
+            "sample_time": (
+                trigger_times[0] + (trigger_times[-1] - trigger_times[0]) / 2
+            ).unix
```
Please don't convert to unix here, just keep it as astropy time object

maxnoe commented on src/ctapipe/calib/camera/extractor.py:
```diff
@@ -0,0 +1,426 @@
+        self.arrival_times = self.arrival_times[self.update_frequency :]
+        self.broken_pixels = self.broken_pixels[self.update_frequency :]
+
+    @abstractmethod
```
private methods should not be abstract

maxnoe commented on src/ctapipe/calib/camera/extractor.py:
```diff
@@ -0,0 +1,426 @@
+        mon.tel[self.tel_id].flatfield container
+        """
+
+        container = event.mon.tel[self.tel_id].flatfield
```
I don't think we should set the new container on the event.

Instead, let the class keep a list of computed calibrations.

maxnoe commented on src/ctapipe/calib/camera/extractor.py:
```diff
@@ -0,0 +1,426 @@
+            self.sample_masked_pixels,
+        )
+
+        flatfield_container = {
```
We made very bad experiences with modifying containers in place. You should create and fill a fresh container.

You can also directly pass the fields to the container itself.

maxnoe commented on src/ctapipe/calib/camera/extractor.py:
```diff
@@ -0,0 +1,426 @@
+        ff_charge_failing_pixels = np.logical_or(
+            container.charge_median_outliers, container.charge_std_outliers
+        )
+        event.mon.tel[self.tel_id].pixel_status.flatfield_failing_pixels = (
```
Again, we shouldn't modify the input event. That's for the code that reads back the calibration coefficients and interpolates / matches them to events when applying the calibration.

maxnoe commented on src/ctapipe/calib/camera/extractor.py:
```diff
@@ -0,0 +1,426 @@
+
+        # append valid dl1 events
+        self.n_events_in_buffer += 1
+        self.trigger_times.append(event.trigger.tel[self.tel_id].time)
```
`self.tel_id` seems not to be defined anymore.

Also: these components should be able to handle multiple telescopes I think, so we keep e.g. a dict from tel_id to the properties.

maxnoe replied:
It also looks like you are implementing a deque with the logic in clean:
https://docs.python.org/3/library/collections.html#collections.deque

E.g. if you always want to keep the most recent N events, use a `deque(maxsize=N)` and just `append` to it.

Once you reach maxsize, the oldest event will be discarded.

maxnoe commented:
@FrancaCassol I think it doesn't make much sense to apply any optimization to the existing code, as we now we need to replace it completely relatively soon.

maxnoe commented:
The logic is already somewhat different to what is implemented in LST currently, so clearly this at least has some new development.

kosack reviewed (commented):
I have a general question:  these classes are for computing running pedestals and flatfields, but they process each event at at time and have mechanisms for bunching events, and keeping running statistics.    Now that we know all calibration events will be in their own file and that for offline processing we can do a first pass on that file to compute coefficents, is the current design with an event-loop really the best one?  The only use case I can see for that is for real-time analysis, but that is not what we have, we can compute all calibration after the run has finished and can even loop many times if needed.

I would think it is much more efficient and easy to implement the computation of calibration coefficients using a procedure that first pre-processes and writes out the events to an HDF5 file, and then using TableLoader, apply functions to all events at once, similar to how `ctapipe-apply-models` works.    Should we maybe have a design discussion about this?  We can still of course integrate this method for the time being, but I suspect it would be much less code and allow fancier algorithms if we process *column-wise* rather than *row-wise*(per event). For example doing a boxcar average for the pedestals for example would be trivial to implement.

TjarkMiener commented:
This PR draft was meant to refactor the current lstchain components for flatfield and pedestal to remove repeated code and start a discussion on the migration. In addition, I added this config parameter, which allows the calculation of the stats after a given iteration and not once the sample buffer is filled. Originally I had in mind to follow what @kosack suggested (pre-process/integrate charges and arrival times, dump dl1-like events, and then table read plus the calculation of coefficients). I dropped this idea because a presumed 2x 100Hz calibration rate and an obs block duration of 20-25 mins would exceed the RAM limitation of 2GB per job when reading the whole table. However, as we discussed in a separate thread, the calibration rate will be likely reduced in the near future. So I guess we could reconsider the suggestion of integrating and temporarily storing dl1-like calibration data to read them afterward using the TableLoader and apply functions to all events at once.

The idea I followed in this PR for the design of the components was to be able to flexibly add subcomponents that could host the "fancy algorithms" via the `_process()` function. You could in principle increase the `sample_size` and then apply the complex functions on the `self.charges` and `self.arrival_times` arrays as you would do for all the events read with the TableLoader. Those complex functions could even be applied on a pixel-wise level, i.e. we might want to treat pixels that are affected by stars differently. The current subcomponents `FlatFieldExtractor` and `PedestalExtractor` were targeting the cat-B processing, while fancy subcomponents can be added in the future for cat-C processing.

I like the idea proposed in #1226 for the on-the-fly computation of flatfield and pedestals, even though we might want to store the charge statistics monitoring of the calibration events for the data quality monitoring.

maxnoe commented:
The table loader also supports chunked loading, which allows you to iterate through chunks and keeping consecutive chunks in memory, you only need roughly the amount of events of your windows sizes.

TjarkMiener commented:
True, good point! My initial plan was to follow a workflow like: integrating charges and extracting arrival times via the `process` tool  `($ ctapipe-process ... --write_images --no-write_parameters)`. And then code a tool for the calculation of the camera calibration coefficients based on the TableLoader. Therefore, I would need to make the `process` tool compatible with calibration events, i.e. allowing two gains integration. By using the `process` tool we would directly ensure to follow the standard `ctapipe` DL1 data model, which then probably needs some adjustments for the allowance of two gains images and peak_times. I think the dl1-like files for the calibration events can be stored temporarily. One advantage would also be that those dl1-like calibration files could be sent directly to the off-site data center after production rather than the waveforms of the calibration events. A naive guess from my side would be that we don't want to repeat the charge integration and arrival time extraction for the calibration events each time we are calculating camera calibration coefficients (on- or off-site).

I'm closing this draft for now and working on the suggestion above in a separate branch. In case you think we should the design of this PR feel free to comment and I will reopen this draft. Thanks for the discussion and suggestion!
